src/hotspot/share/prims/jvmtiEnvBase.cpp

```diff
@@ -756,6 +756,17 @@ JvmtiEnvBase::get_thread_state(oop thread_oop, JavaThread* jt) {
   return state;
 }
 
+jint
+JvmtiEnvBase::get_thread_state(oop thread_oop, JavaThread* jt) {
+  jint state = get_thread_state_base(thread_oop, jt);
+
+  if (is_thread_carrying_vthread(jt, thread_oop)) {
+    state &= ~JVMTI_THREAD_STATE_RUNNABLE;
+    state |= JVMTI_THREAD_STATE_WAITING | JVMTI_THREAD_STATE_WAITING_INDEFINITELY;
```

Comment:
This does not look correct.
GetThreadState spec provides hierarchical set of questions to interpret thread state value.
JVMTI_THREAD_STATE_ALIVE | JVMTI_THREAD_STATE_WAITING | JVMTI_THREAD_STATE_WAITING_INDEFINITELY is only one branch and I'd expect all other bits are not set for this state.
Need to decide what do we want to report as carrier thread state for all possible values returned by get_thread_state_base().

Reply:
Good concern.
There are two bits (and the related RUNNABLE bit) that we care in this sub-tree of state bits: `SUSPENDED` and `INTERRUPTED`. This update clones these two bits. The RUNNABLE bit must be cleared.
A thread carrying a virtual thread can not be in native, blocked, parked, sleeping or waiting on some object.
The state returned by the `get_thread_state_base` is based on the call:
`    state = (jint)java_lang_Thread::get_thread_status(thread_oop);`
and addition of the derived from JavaThread bits: `SUSPENDED`, `INTERRUPTED` and `IN_NATIVE`.
The three bit derived from the JavaThread are not relevant.
This call has to be made directly:
`   state = (jint)java_lang_Thread::get_thread_status(thread_oop);`
The SUSPEND bit has to be based on the call:
`    java_thread->is_carrier_thread_suspended();`

Reply:
Do you need to check jt->is_interrupted(false) and set INTERRUPTED bit?
It looks like java_lang_Thread::get_thread_status(thread_oop) can only return RUNNABLE in the case and we clear it.

    if (is_thread_carrying_vthread(jt, thread_oop)) {
      jint state = JVMTI_THREAD_STATE_WAITING | JVMTI_THREAD_STATE_WAITING_INDEFINITELY;
      if (jt->is_carrier_thread_suspended()) {
        state |= JVMTI_THREAD_STATE_SUSPENDED;
      }
      if (jt->is_interrupted(false)) {
        state |= JVMTI_THREAD_STATE_INTERRUPTED;
      }
      return state;
    } else ...

Reply:
> A thread carrying a virtual thread can not be in native, blocked, parked, sleeping or waiting on some object.

Actually it can be in native.
And if I remember correctly synchronized block pins virtual thread, so inside synchronized we can get other states

Reply:
The INTERRUPTED bit we need has to be returned by the `java_lang_Thread::get_thread_status`.
Not completely sure but the bit jt->is_interrupted(false) can be set for the mounted virtual thread.
The JVMTI InterruptThread calls this function to set interrupt bit for non-virtual threads:
`  java_lang_Thread::set_interrupted(thread_obj, true);`

Reply:
Corrected the function `get_thread_state()` to make it more safe.
Only `ALIVE` and `INTERRUPTED` bits are taken from result of `java_lang_Thread::get_thread_status(thread_oop)`.

Reply:
> A thread carrying a virtual thread can not be in native, blocked, parked, sleeping or waiting on some object.

A virtual thread can call native code, be blocked on an object monitor, or waiting on an object monitor. Only parking and sleeping are specialized for virtual threads in the list you gave.

Reply:
> > A thread carrying a virtual thread can not be in native, blocked, parked, sleeping or waiting on some object.

>A virtual thread can call native code, be blocked on an object monitor, or waiting on an object monitor. Only parking and sleeping are specialized for virtual threads in the list you gave.

This statement was about carrier thread when there is a virtual thread executed at the top.
We are getting state bits with the `java_lang_Thread::get_thread_status(thread_oop)` where the `thread_oop` belongs to the carrier thread. But you are talking about a virtual thread which, of course, can be in almost any state.

Reply:
Thanks for clarifying - it gets very confusing as to which "thread" is being talked about. But if a virtual thread is mounted on this JavaThread then I thought the carrier thread's thread-oop is supposed to be in a blocked state?

Reply:
It was decided with Alan that it is okay to be waiting state.
Alan's comment in the original PR [https://github.com/openjdk/jdk/pull/14298](https://github.com/openjdk/jdk/pull/14298) was:
>  if the jt is carrying thread_oop and it's okay for the JVMTI state to reported as WAITING when waiting for something other than Object.wait.

Reply:
The mental model  is that the carrier is blocked so this is what an observer using the APIs should see. My recollection is that JVMTI_THREAD_STATE_WAITING was okay because there is a wriggle room in the JVM TI spec, it only uses Object.wait as an example. There may be a few rough edges to smooth down in this area. It's okay to take time with this PR and expand the tests to cover more cases and get more confident that there aren't more issues.

Reply:
We agreed with Alex to file a test RFE to improve test coverage in this area.